freqtrade/exchange/binance_public_data.py

```diff
@@ -0,0 +1,216 @@
+
+from freqtrade.enums import CandleType
+from freqtrade.misc import chunks
+from freqtrade.util.datetime_helpers import dt_from_ts, dt_now
+
+
+logger = logging.getLogger(__name__)
+
+
+class BadHttpStatus(Exception):
+    """Not 200/404"""
+
+    pass
+
+
+async def fetch_ohlcv(
+    candle_type: CandleType,
+    pair: str,
+    timeframe: str,
+    since_ms: int,
+    until_ms: int | None,
+    stop_on_404: bool = False,
+) -> DataFrame:
+    """
+    Fetch OHLCV data from https://data.binance.vision/
+    :candle_type: Currently only spot and futures are supported
+    :param until_ms: `None` indicates the timestamp of the latest available data
+    :param stop_on_404: Stop to download the following data when a 404 returned
+    :return: the date range is between [since_ms, until_ms),
+        return and empty DataFrame if no data available in the time range
+    """
+    try:
+        if candle_type == CandleType.SPOT:
+            asset_type = "spot"
+        elif candle_type == CandleType.FUTURES:
+            asset_type = "futures/um"
+        else:
+            raise ValueError(f"Unsupported CandleType: {candle_type}")
+        symbol = symbol_ccxt_to_binance(pair)
+        start = dt_from_ts(since_ms)
+        end = dt_from_ts(until_ms) if until_ms else dt_now()
+
+        # We use two days ago as the last available day because the daily archives are daily
+        # uploaded and have several hours delay
+        last_available_date = dt_now() - datetime.timedelta(days=2)
+        end = min(end, last_available_date)
+        if start >= end:
+            return DataFrame()
+        df = await _fetch_ohlcv(asset_type, symbol, timeframe, start, end, stop_on_404)
+        logger.debug(
+            f"Downloaded data for {pair} from https://data.binance.vision with length {len(df)}."
+        )
+    except Exception as e:
+        logger.debug("An exception occurred", exc_info=e)
+        df = DataFrame()
+
+    if not df.empty:
+        return df.loc[(df["date"] >= start) & (df["date"] < end)]
+    else:
+        return df
+
+
+def symbol_ccxt_to_binance(symbol: str) -> str:
+    """
+    Convert ccxt symbol notation to binance notation
+    e.g. BTC/USDT -> BTCUSDT, BTC/USDT:USDT -> BTCUSDT
+    """
+    if ":" in symbol:
+        parts = symbol.split(":")
+        if len(parts) != 2:
+            raise ValueError(f"Cannot recognize symbol: {symbol}")
+        return parts[0].replace("/", "")
+    else:
+        return symbol.replace("/", "")
+
+
+def concat(dfs) -> DataFrame:
+    if all(df is None for df in dfs):
+        return DataFrame()
+    else:
+        return pd.concat(dfs)
+
+
+async def _fetch_ohlcv(
+    asset_type: str,
+    symbol: str,
+    timeframe: str,
+    start: datetime.date,
+    end: datetime.date,
+    stop_on_404: bool,
+) -> DataFrame:
+    dfs: list[DataFrame | None] = []
+
+    connector = aiohttp.TCPConnector(limit=100)
+    async with aiohttp.ClientSession(connector=connector) as session:
```

Comment:
How will this handle proxies? 
in dream mode - it'd take it from the ccxt_config.httpsProxy object (https://www.freqtrade.io/en/latest/configuration/#using-a-proxy-with-freqtrade) - but i'm not sure if that's a good idea as this can really be httpProxy, socksProxy, (ccxt documentation has them all).

the alternative is to use the env variables ... maybe that's builtin - but i didn't really check for this.

Reply:
Using the environment variables seems more correct, however, it might be confusing for users why we need two proxies for download binance candlesticks 😅

Currently use  enviroment variables:
[feat: support proxy from environment variables](https://github.com/freqtrade/freqtrade/pull/10858/commits/d7555e1f2971b6bcbbd4d978bd0a81e436c91f3d)